src/common/stack.rs

```diff
@@ -1,88 +1,419 @@
+	/// 	StackSize::from_element_count(1024).into_initial(),
+	/// 	StackSize::unlimited(),
+	/// );
+	/// ```
+	///
+	/// # Errors
+	///
+	/// In debug mode, panics if `initial_size` is larger than `limit`.
+	pub fn new(initial_size: StackSizeInitial<T>, limit: StackSizeLimit<T>) -> StackWithLimit<T> {
+		let initial_size_elements = initial_size.0.element_count();
+		let limit_elements = limit.0.element_count();
+		debug_assert!(
+			limit_elements >= initial_size_elements,
+			"initial_size should not be larger than StackWithLimit limit"
+		);
+		StackWithLimit {
+			stack: Vec::with_capacity(initial_size_elements.min(limit_elements)),
+			limit: limit_elements,
+		}
+	}
 
-impl fmt::Display for Error {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(f, "{}", self.0)
+	/// Create an new StackWithLimit with `limit` max size and `limit` elements pre-allocated
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::{StackWithLimit, StackSize};
+	/// let bstack = StackWithLimit::<i32>::with_size(StackSize::from_element_count(2));
+	/// ```
+	pub fn with_size(size: StackSize<T>) -> StackWithLimit<T> {
+		StackWithLimit::new(StackSizeInitial(size), StackSizeLimit(size))
+	}
+
+	/// Attempt to push value onto stack.
+	///
+	/// # Errors
+	///
+	/// Returns Err(StackOverflow) if stack is already full.
+	#[inline]
+	pub(crate) fn push(&mut self, value: T) -> Result<(), StackOverflow> {
+		if self.stack.len() == self.stack.capacity() {
+			if self.stack.len() == self.limit {
+				return Err(StackOverflow);
+			}
+			// grows exponentially, just like Vec normally does
+			// https://doc.rust-lang.org/1.26.0/src/alloc/raw_vec.rs.html#462
+			let desired_len = self
+				.stack
+				.len()
+				.checked_mul(2)
+				.unwrap_or(usize::MAX)
+				.min(self.limit)
+				.max(1);
+			let additional_len = desired_len - self.stack.len();
+			self.stack.reserve_exact(additional_len);
+		}
+		debug_assert!(self.stack.len() < self.limit);
+		debug_assert!(self.stack.len() < self.stack.capacity());
+		let len = self.stack.len();
+		unsafe {
+			ptr::write(self.stack.get_unchecked_mut(len), value);
+			self.stack.set_len(len + 1);
+		}
+		Ok(())
+	}
+
+	pub(crate) fn pop(&mut self) -> Option<T> {
+		self.stack.pop()
+	}
+
+	/// Remove and Return top element. Does not check for emptyness.
+	/// If this is called on a zero length stack, bad things will happen.
+	/// Do not call this method unless you can prove the stack has length.
+	#[inline]
+	pub(crate) unsafe fn pop_unchecked(&mut self) -> T {
+		debug_assert!(self.stack.len() > 0);
+		let len = self.stack.len();
+		self.stack.set_len(len - 1);
+		ptr::read(self.stack.get_unchecked(self.stack.len()))
+	}
+
+	/// Return optional reference to item `depth` distance away from top
+	///
+	/// `bstack.get_relative_to_top(0)` gets the top of the stack
+	///
+	/// `bstack.get_relative_to_top(1)` gets the item just below the stack
+	pub(crate) fn get_relative_to_top(&self, depth: usize) -> Option<&T> {
```

Comment:
Generally Rust avoids `get_noun`-format names. I would call this `nth_from_top`.